### ISSUE TITLE
[Question] Attach Filter() to observable data stream in different class

### ISSUE THREAD

**Issue by PlotlyUser2** [reactiveui/DynamicData#340]
I'm totally new to the reactive extensions / reactive ui and dynamic data stuff in C#, so I therefore apologise in advance if my question is not that clear...

I'm facing currently a following problem:

I have to views and therefore two Viewmodels. Now, I try to attach the **Filter()** method within another class while the original connection to an **IObservableList<T> SourceList** is created in a different class.

Example snippets:

**Class 1**

```
    // sourceList is used to fill the ReadOnlyObservableCollection<T> while receiving data from an event pattern

    sourceList.Connect()               // SourceList<T>
              .Bind(out _itemsBinding) // private ReadOnlyObservableCollection<T>
              .DisposeMany()
              .Subscribe();
```

**Class 2**

```
        var observableSearchPhrase = this
            .WhenAnyValue(x => x.FilterExpression) // [Reactive] Property of type string
            .Where(x => !string.IsNullOrWhiteSpace(x) && !string.IsNullOrEmpty(x))
            .Select(FilterOnId);

        sourceList.Connect().Filter(observableSearchPhrase).Subscribe(); //SourceList<T> instance from Class 1

```
**The predicate method in Class 2**

    private Func<T, bool> FilterOnId(string id)
    {
        return n => n.Id >= int.Parse(id);
    }

When I run the application, I don't see any invocation of the filter on to my data stream, so the line:

    sourceList.Connect().Filter(observableSearchPhrase).Subscribe(); //SourceList<T> instance from Class 1

does nothing. So basically, my approach was to reuse a data stream, but attach a Filter dynamically based on the **FilterExpression** input.

How can it be fixed? Does my approach even points into the right direction?

I've viewed the dynamicfilter example by dynamic data, but it is always used in the same class, where the predicate filter function is also written.

**Comment by Claye-L** [comment]
The reason your thing doesnt work is simple. The Filter() operator (like most other dynamicdata operators) doesnt actually modify the sourceCache. Instead it creates a new collection that has been filtered. Basically you need to think of dynamicdata operators as being a one way pipeline.
So to fix your problem, you need to actually write the Filter() before the Bind() in class1.
You will need to reference class2 from class1.
I understand that your situtation is you have a view+viewmodel with data to display (class1) and a view+viewmodel with a filter for the first page (class2).

When i'm in this case i do the following:

class2:
```cs
    private IObservable<Func<T,bool>> Filter =>this
            .WhenAnyValue(x => x.FilterExpression)
            .Where(x => !string.IsNullOrWhiteSpace(x) && !string.IsNullOrEmpty(x))
            .Select(FilterOnId);
```

class1:
```cs
private class2 filterViewModel; //TODO get this injected using your dependency injection of choice

    sourceList.Connect()
           .Filter(filterViewModel.Filter) // the filter needs to be before the bind because the filter only affects things that come after it (or downstream)
           .Bind(out _itemsBinding) 
           .DisposeMany()
           .Subscribe();
```

This is pretty much the only way to do what you want.

**Comment by PlotlyUser2** [comment]
@Claye-L Thank you for your explanation! I somehow missed the point, that the **Filter()** creates a new filtered collection instead of modifying the sourcelist. As you pointed out, your solution would be the only one that I can use.

> I understand that your situtation is you have a view+viewmodel with data to display (class1) and a view+viewmodel with a filter for the first page (class2).

Yes, that's right. I have a datagrid with data content (class1) and a filter section (class2). 

However, in my case, I've multiple filters (imaging having multiple searchboxes), where each filter can affect the columns (or two/three columns at the same time) on the datagrid. 

So basically, I need to build something like this: http://www.albahari.com/nutshell/predicatebuilder.aspx and apply it to the **.Filter(filterPredicate.Filter)**.

Would it be the right approach? Because of the possible filter combinations, I simply can not write **.Filter(filterViewModel.Filter1).Filter(filterViewModel.Filter2).Filter(filterViewModel.FilterX)** or, can I inject them somehow within the data stream pipeline?

It would be nice, if the dynamic data would have something like this `<IEnumerable<IObservable<Func<T,bool>>>` where each Filter is applied on the sourcelist with an **Or** condition.

**Comment by Claye-L** [comment]
you want a function from IEnumerable<Func<T,bool>> to Func<T,bool>. Booleans are a monoid with respect to AND, so you can just use the LINQ Aggregate operator on this.

so Say you have many viewmodels class2 that do different filters and have 
```cs
public IObservable<Func<T,bool>> Filter1 {get;} //implementation hidden
public IObservable<Func<T,bool>> Filter2 {get;}
public IObservable<Func<T,bool>> Filter3 {get;}
public IObservable<Func<T,bool>> FilterX {get;}

public IObservable<Func<T,bool>> AllFiltersCombined => Filter1.CombineLatest(Filter2,Filter3,FilterX, (f1,f2,f3,fx) => AggregatePredicatesAND(f1,f2,f3,fx));

//using this helper method
//takes an IEnumerable of predicates, and returns a single predicate that applies all filters to an item
//which means, the new predicate returns true if an item verifies ALL filters
public static Func<T,Bool> AggregatePredicatesAND(params Func<T,bool>[] predicates) 
        {
            return predicates.Aggregate<Func<T,bool>>((fa,fb) => (T t) => fa(t) && fb(t));
        }
```
in your class1 just filter once on this new AllFiltersCombined property
